### PR TITLE
Cache manager and cache invalidation

### PR DESCRIPTION
**Pull request by Fnuworsu**
# Description  
- What does this PR do?  
   `Implements a comprehensive in-memory caching system with middleware and invalidation strategies for learning hubs, subhubs, users, and sessions`  
- Why is it necessary or valuable?  
  `Reduces database load and API response times by caching frequently accessed data, with specialized caches for different entity types and automatic invalidation`  
- What is intentionally left out and will be done in a later PR?  
  `Distributed cache support, cache persistence across server restarts, and cache analytics dashboard`

---

# Milestones / Related Work  
- Feature or user story this contributes to: `Performance Optimization Initiative`  
---

# Resources and References  
- Tutorials, documentation, or code examples used:  
  `Node.js Map documentation, Express middleware patterns`  
- Any copied or adapted code (with attribution):  
  `<Mention if any logic was reused or adapted>`  
---

# Test Plan  
- How was this tested?  
  `Manual testing with API endpoints to verify cache hits/misses and invalidation patterns  

## Edge Cases Tested  
*   `Cache invalidation cascades across related entities`
*   `Concurrent access patterns`
*   `Memory limit enforcement with LRU eviction`
*   `Error handling for all public methods`
*   `Missing/invalid IDs handling`

---

# Additional Notes  
- Known limitations: ``Memory usage calculation is approximate and doesn't account for JavaScript object overhead``  
- Planned follow-up PRs: `Cache analytics dashboard, distributed cache support with Redis`  
- Reviewer requests or feedback needed: `Looking for feedback on the cache invalidation strategy and middleware implementation`

### CODE REVIEW

**Comment by Fnuworsu** [comment]
@sinap-fb Merged other sub cache branches here

File: backend/cache/cacheManager.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+               maxSize: 200,
+               defaultTTL: 1800000
+           })
+           this.sessionCache = new InMemoryCache({
+               maxSize: 1000,
+               defaultTTL: 3600000
+           })
+           CacheManager.instance = this
+       } catch (error) {
+           throw new Error(`Failed to initialize CacheManager: ${error.message}`)
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/cache/cacheManager.js:22]: @sinap-fb had a quick read through the mdn docs you sent. Looking through the examples, I see why yiou suggested it!

File: backend/routes/cacheRoutes/cacheRoutes.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
+const express = require("express")
+const cacheManager = require("../../cache/cacheManager")
+const SimpleInvalidator = require("../../cache/simpleInvalidator")
+const { isAuthenticated } = require("../../middleware/middleware.js")
+const router = express.Router()
+const invalidator = new SimpleInvalidator(cacheManager)
+
+
+// get cache statistics
+router.get("/api/admin/cache/stats", isAuthenticated, (req, res) => {
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/routes/cacheRoutes/cacheRoutes.js:10]: @sinap-fb appreciate it!

File: backend/routes/cacheRoutes/cacheRoutes.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+})
+
+
+// clear cache for specific user
+router.post("/api/admin/cache/clear-user", isAuthenticated, async (req, res) => {
+   try {
+       const { userId } = req.body
+
+       if (!userId) {
+           return res.status(400).json({
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/routes/cacheRoutes/cacheRoutes.js:66]: thanks @sinap-fb

File: backend/cache/cacheManager.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+   async getCachedUserHubs(userId) {
+       try {
+           if (!userId) throw new Error('User ID is required')
+           return this.learningHubCache.getUserHubs(userId)
+       } catch (error) {
+           throw new Error(`Failed to get cached user hubs: ${error.message}`)
+       }
+   }
+
+   async invalidateUserHubs(userId) {
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/cache/cacheManager.js:45]: I initially implemented cache persistence using Prisma to maintain cache state across server restarts, which required async database operations. However, I realized this defeated the core purpose of in-memory caching - eliminating database round trips. Persisting cache operations back to the database created a write-amplification problem where every cache invalidation triggered additional database writes, negating the performance benefits(read this from designing data intensive applications btw).

[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/cache/cacheManager.js:45]: @sinap-fb I made the changes.

File: backend/cache/simpleInvalidator.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        try {
+            if (!userId) throw new Error('User ID is required for cache invalidation')
+            // clear all user-related caches in parallel
+            const invalidationTasks = [
+                this.cacheManager.invalidateUserHubs(userId),
+                this.cacheManager.subHubCache.invalidateUserSubhubs(userId),
+                this.cacheManager.userCache.delete(`user:${userId}`),
+                this.cacheManager.userCache.delete(`user:${userId}:profile`)
+            ]
+            await Promise.all(invalidationTasks)
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/cache/simpleInvalidator.js:16]: @sinap-fb I made the changes.

File: backend/middleware/cacheMiddleware.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+const cacheMiddleware = (keyGenerator, ttl = 600000) => {
+   return async (req, res, next) => {
+       try {
+           const cacheKey = keyGenerator(req)
+           const cachedData = await cacheManager.userCache.get(cacheKey)
+
+           if (cachedData) return res.json(cachedData)
+           const originalJson = res.json
+           // override res.json to cache the response
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/middleware/cacheMiddleware.js:11]: i could create a cache helper to res object
```javascript
res.cacheAndSend = (data) => {
    cacheManager.userCache.set(cacheKey, data, ttl)
    return res.json(data)
}
```

File: backend/cache/cacheManager.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (sinap-fb) [reply backend/cache/cacheManager.js:45]: ooh nice finding!
